common/txmgr/inmemory_store.go

```diff
@@ -0,0 +1,1869 @@
+	for _, as := range ms.addressStates {
+		as.ApplyToTxsByState(nil, fn)
+	}
+
+	return nil
+}
+
+// FindTxAttemptsConfirmedMissingReceipt returns all transactions that are confirmed but missing a receipt
+func (ms *InMemoryStore[CHAIN_ID, ADDR, TX_HASH, BLOCK_HASH, R, SEQ, FEE]) FindTxAttemptsConfirmedMissingReceipt(ctx context.Context, chainID CHAIN_ID) (
+	[]txmgrtypes.TxAttempt[CHAIN_ID, ADDR, TX_HASH, BLOCK_HASH, SEQ, FEE],
+	error,
+) {
+	if ms.chainID.String() != chainID.String() {
+		return nil, fmt.Errorf("find_next_unstarted_transaction_from_address: %w", ErrInvalidChainID)
+	}
+
+	filter := func(tx *txmgrtypes.Tx[CHAIN_ID, ADDR, TX_HASH, BLOCK_HASH, SEQ, FEE]) bool {
+		return tx.TxAttempts != nil && len(tx.TxAttempts) > 0
+	}
+	states := []txmgrtypes.TxState{TxConfirmedMissingReceipt}
+	attempts := []txmgrtypes.TxAttempt[CHAIN_ID, ADDR, TX_HASH, BLOCK_HASH, SEQ, FEE]{}
+	ms.addressStatesLock.RLock()
+	defer ms.addressStatesLock.RUnlock()
+	for _, as := range ms.addressStates {
+		attempts = append(attempts, as.FetchTxAttempts(states, filter)...)
+	}
+	// sort by tx_id ASC, gas_price DESC, gas_tip_cap DESC
+	sort.SliceStable(attempts, func(i, j int) bool {
+		return attempts[i].TxID < attempts[j].TxID
+	})
+
+	// deep copy the attempts
+	var eAttempts []txmgrtypes.TxAttempt[CHAIN_ID, ADDR, TX_HASH, BLOCK_HASH, SEQ, FEE]
+	for _, attempt := range attempts {
+		eAttempts = append(eAttempts, ms.deepCopyTxAttempt(attempt.Tx, attempt))
+	}
+
+	return eAttempts, nil
+}
+
+// UpdateBroadcastAts updates the broadcast_at time for a given set of attempts
+func (ms *InMemoryStore[CHAIN_ID, ADDR, TX_HASH, BLOCK_HASH, R, SEQ, FEE]) UpdateBroadcastAts(ctx context.Context, now time.Time, txIDs []int64) error {
+	// Persist to persistent storage
+	if err := ms.txStore.UpdateBroadcastAts(ctx, now, txIDs); err != nil {
+		return err
+	}
+
+	// Update in memory store
+	fn := func(tx *txmgrtypes.Tx[CHAIN_ID, ADDR, TX_HASH, BLOCK_HASH, SEQ, FEE]) {
+		if tx.BroadcastAt != nil && tx.BroadcastAt.Before(now) {
+			tx.BroadcastAt = &now
+		}
+	}
+
+	ms.addressStatesLock.RLock()
+	defer ms.addressStatesLock.RUnlock()
+	for _, as := range ms.addressStates {
+		as.ApplyToTxsByState(nil, fn, txIDs...)
+	}
+
+	return nil
+}
+
+// UpdateTxsUnconfirmed updates the unconfirmed transactions for a given set of ids
+func (ms *InMemoryStore[CHAIN_ID, ADDR, TX_HASH, BLOCK_HASH, R, SEQ, FEE]) UpdateTxsUnconfirmed(ctx context.Context, txIDs []int64) error {
+	// Persist to persistent storage
+	if err := ms.txStore.UpdateTxsUnconfirmed(ctx, txIDs); err != nil {
+		return err
+	}
+
+	// Update in memory store
+	ms.addressStatesLock.RLock()
+	defer ms.addressStatesLock.RUnlock()
+	wg := sync.WaitGroup{}
+	for _, as := range ms.addressStates {
+		wg.Add(1)
+		go func(as *AddressState[CHAIN_ID, ADDR, TX_HASH, BLOCK_HASH, R, SEQ, FEE]) {
+			for _, txID := range txIDs {
+				if err := as.MoveConfirmedMissingReceiptToUnconfirmed(txID); err != nil {
```

Comment:
I think this method is assuming some logic from the method is function is called from. I noticed `UpdateTxsUnconfirmed` is currently only called in the context of moving txs from `TxConfirmedMissingReceipt` to `TxUnconfirmed` but the memory store function to mark tx as unconfirmed only cares about txIDs. To maintain existing behavior, think we should apply this change across all tx states as long as their IDs match.